core/embed/boardloader/main.c

```diff
@@ -276,6 +276,16 @@ int main(void) {
   display_init();
   display_clear();
 
+#ifdef STM32U5
+  const image_header *old_hdr = read_image_header(
+      (const uint8_t *)BOOTLOADER_START, BOOTLOADER_IMAGE_MAGIC,
+      flash_area_get_size(&BOOTLOADER_AREA));
+  secret_ensure_initialized();
+  if (old_hdr != NULL) {
+    get_and_fill_bootloader_version(old_hdr->monotonic);
```

Comment:
I believe that `copy_sd` should not allow installing a downgrade in the first place, i.e., the check should be (a) in `copy_sd` and (b) when getting ready to jump to bootloader.

Placing the check here seems weird. What I would expect the most is splitting `get_and_fill` into `get` and `fill`, with `get` performing the check & halt, and `fill` just writing data. Then `copy_sd` would call `get` (and possibly halt), while the finalization would call both.

Reply:
This actually does not do the check, just ensuring that up-to date version if written before doing the SD card update. Which i think i needed for some unlikely scenarios the stored version gets deleted, or if for example firmware updates bootloader, trezor reboots with SD card containing old bootloader plugged in - which would otherwise result in downgrading the bootloader again.

Otherwise, divided the function in https://github.com/trezor/trezor-firmware/commit/e9f84176037ef83e7fefee99e2c20586e8dd6926 , also the check in SD update is added. `Get` function does not do the halt immediately, leaves it to the caller.